### PR TITLE
feat: add HTTP Gateway support for /ipns/ paths 

### PR DESCRIPTION
**Pull request by lidel**
> Part of an effort to run embedded js-ipfs in Brave :lion: https://github.com/ipfs-shipyard/ipfs-companion/issues/716
> Fixes https://github.com/ipfs/js-ipfs/issues/1918 

This PR will add support for `/ipns/` paths at HTTP Gateway.
Smoke test: [/ipns/tr.wikipedia-on-ipfs.org](http://127.0.0.1:9090/ipns/tr.wikipedia-on-ipfs.org/wiki/Anasayfa.html) (IPNS+DNSLink+HAMT-sharded website)

## Do not merge it yet
 
This PR requires below PRs to be merged first:

  - [x] Gateway Improvements from https://github.com/ipfs/js-ipfs/pull/1989
    (after merging #1989 I will rebase this PR, which will remove first two commits)
  - [x] PeerID 
    eg. `/ipns/<PeerId-as-multihash-b58>`
    - requires https://github.com/ipfs/js-ipfs/pull/2002 to land first
  - [x] `/ipns/<libp2p-key-in-cidv1>` 
    - requires https://github.com/multiformats/js-multicodec/pull/45
  - [x] DNSLink 
    eg. `/ipns/<fqdn>/path/file` like `/ipns/docs.ipfs.io/assets/logo.svg`
    - requires https://github.com/ipfs/js-ipfs/pull/2002 to land first
  - [x] HAMT shard support 
    eg. `/ipns/tr.wikipedia-on-ipfs.org/wiki/Anasayfa.html` (`wiki` is a sharded directory)
    - requires https://github.com/ipfs/js-ipfs-http-response/issues/22 and https://github.com/ipfs/js-ipfs-mfs/pull/48  to land first
  - [x] Tests for `/ipns/`

### CODE REVIEW

**Comment by lidel** [comment]
I rebased and switched to released versions of `ipfs-mfs` and `ipfs-http-response`.
Test link for HTTP Gateway (IPNS+DNSLink+HAMT):

- http://127.0.0.1:9090/ipns/tr.wikipedia-on-ipfs.org/wiki/Anasayfa.html

I resolve `/ipns/` to `/ipfs/` before passing it to `ipfs-http-response` as a workaround until  https://github.com/ipfs/js-ipfs-http-response/issues/22 is solve.

File: src/http/gateway/routes/index.js [hunk 1]
```diff
@@ -1,3 +1,3 @@
 'use strict'
 
-module.exports = [require('./gateway')]
+module.exports = [...require('./gateway')]
```
Comment (hugomrdias) [inline src/http/gateway/routes/index.js]: you dont need to spread here the default exports is alread an array
```suggestion
module.exports = require('./gateway')
```

File: package.json [hunk 2]
```diff
@@ -94,8 +94,8 @@
     "ipfs-block": "~0.8.1",
     "ipfs-block-service": "~0.15.1",
     "ipfs-http-client": "^32.0.1",
-    "ipfs-http-response": "~0.3.0",
-    "ipfs-mfs": "~0.11.4",
+    "ipfs-http-response": "~0.3.1",
+    "ipfs-mfs": "~0.11.5",
```
Comment (hugomrdias) [inline package.json]:
```suggestion

```
its not used

File: package.json [hunk 3]
```diff
@@ -94,8 +94,8 @@
     "ipfs-block": "~0.8.1",
     "ipfs-block-service": "~0.15.1",
     "ipfs-http-client": "^32.0.1",
-    "ipfs-http-response": "~0.3.0",
-    "ipfs-mfs": "~0.11.4",
+    "ipfs-http-response": "~0.3.1",
```
Comment (hugomrdias) [inline package.json]:
```suggestion
```
its not used

File: src/http/gateway/resources/gateway.js [hunk 4]
```diff
@@ -45,29 +51,40 @@ class ResponseStream extends PassThrough {
...
-    if (!request.params.cid) {
+  checkImmutableId (request, h) {
+    if (!request.params.immutableId) {
       throw Boom.badRequest('Path Resolve error: path must contain at least one component')
     }
-
-    return { ref: `/ipfs/${request.params.cid}` }
+    return { ref: `/ipfs/${request.params.immutableId}` }
+  },
+  checkMutableId (request, h) {
```
Comment (hugomrdias) [inline src/http/gateway/resources/gateway.js]: can we make this check with https://hapijs.com/tutorials/validation ?

File: src/http/gateway/resources/gateway.js [hunk 5]
```diff
@@ -45,29 +51,40 @@ class ResponseStream extends PassThrough {
 }
 
 module.exports = {
-  checkCID (request, h) {
-    if (!request.params.cid) {
+  checkImmutableId (request, h) {
```
Comment (hugomrdias) [inline src/http/gateway/resources/gateway.js]: can we make this check with https://hapijs.com/tutorials/validation ?

File: src/http/gateway/routes/gateway.js [hunk 6]
```diff
@@ -2,19 +2,37 @@
...
-    ],
-    response: {
-      ranges: false // disable built-in support, we do it manually
-    },
-    ext: {
-      onPostHandler: { method: resources.gateway.afterHandler }
+module.exports = [
+  {
+    method: '*',
+    path: '/ipfs/{immutableId*}',
```
Comment (hugomrdias) [inline src/http/gateway/routes/gateway.js]: i think we should keep this as CID or else we are just making up another concept and we already have enough

File: src/http/gateway/routes/gateway.js [hunk 7]
```diff
@@ -2,19 +2,37 @@
...
+        ranges: false // disable built-in support, we do it manually
+      },
+      ext: {
+        onPostHandler: { method: resources.gateway.afterHandler }
+      }
+    }
+  },
+  {
+    method: '*',
+    path: '/ipns/{mutableId*}',
```
Comment (hugomrdias) [inline src/http/gateway/routes/gateway.js]: we should define what we should call this id, libp2p-key, ipns-key-id, even ipns-cid (content been the hash of the pubkey) and we should PR to the spec https://github.com/ipfs/specs/tree/master/naming this new name.

File: src/http/gateway/routes/gateway.js (hunk 6 above)
Comment (lidel) [reply src/http/gateway/routes/gateway.js]: I looked closer and this path is actually cid+optional path, so I renamed it to `cidPath` (name already used in [one of is-ipfs checks](https://www.npmjs.com/package/is-ipfs#isipfscidpathpath)).

File: src/http/gateway/routes/gateway.js (hunk 7 above)
Comment (lidel) [reply src/http/gateway/routes/gateway.js]: We have [libp2p-key multicodec now](https://github.com/multiformats/js-multicodec/pull/45) but it also can be FQDN with DNSLink, so I renamed it to `libp2pKeyOrFqdn`

File: src/http/gateway/routes/gateway.js (hunk 6 above)
Comment (hugomrdias) [reply src/http/gateway/routes/gateway.js]: what about using `ipfs path` and `ipns path` both represent more than cid/keys

so a string representation of this type `/ipfs/<something>*` would be ipfs path, same for ipns 


CID is an object with a string representation as a baseX of the multihash.

Comment (lidel) [reply src/http/gateway/routes/gateway.js]: To keep things simple I renamed both to `ipfsPath` and `ipnsPath` respectively: https://github.com/ipfs/js-ipfs/pull/2020/commits/eebde0bc8591f63066e665d65f81c8646a15e083
Maybe it should be just `path` in both? I really want to remove custom names where possible.

File: src/http/gateway/resources/gateway.js [hunk 8]
```diff
@@ -45,29 +53,30 @@ class ResponseStream extends PassThrough {
...
   async handler (request, h) {
-    const { ref } = request.pre.args
+    const ref = request.path
     const { ipfs } = request.server.app
 
+    // The resolver from ipfs-http-response supports only immutable /ipfs/ for now,
+    // so we convert /ipns/ to /ipfs/ before passing it to the resolver ¯\_(ツ)_/¯
+    // This could be removed if a solution proposed in
+    //  https://github.com/ipfs/js-ipfs-http-response/issues/22 lands upstream
+    const immutableRef = decodeURIComponent(ref.startsWith('/ipns/')
```
Comment (hugomrdias) [inline src/http/gateway/resources/gateway.js]:
```suggestion
    const ipfsPath = decodeURIComponent(ref.startsWith('/ipns/')
```

File: src/http/gateway/resources/gateway.js [hunk 9]
```diff
@@ -45,29 +53,30 @@ class ResponseStream extends PassThrough {
...
+      ? await resolveIpns(ref, ipfs)
+      : ref)
+
     let data
     try {
-      data = await resolver.cid(ipfs, ref)
+      data = await resolver.cid(ipfs, immutableRef)
```
Comment (hugomrdias) [inline src/http/gateway/resources/gateway.js]:
```suggestion
      data = await resolver.cid(ipfs, ipfsPath)
```

File: src/http/gateway/resources/gateway.js [hunk 10]
```diff
@@ -45,29 +53,30 @@ class ResponseStream extends PassThrough {
...
     } catch (err) {
       const errorToString = err.toString()
       log.error('err: ', errorToString, ' fileName: ', err.fileName)
 
       // switch case with true feels so wrong.
       switch (true) {
         case (errorToString === 'Error: This dag node is a directory'):
-          data = await resolver.directory(ipfs, ref, err.cid)
+          data = await resolver.directory(ipfs, immutableRef, err.cid)
```
Comment (hugomrdias) [inline src/http/gateway/resources/gateway.js]:
```suggestion
          data = await resolver.directory(ipfs, ipfsPath, err.cid)
```

File: src/http/gateway/routes/gateway.js (hunk 6 above)
Comment (hugomrdias) [reply src/http/gateway/routes/gateway.js]: yes we should just call this `path` 

ipfsPath would be the complete request.path `'/ipfs/{path*}'`

File: src/http/gateway/resources/gateway.js [hunk 11]
```diff
@@ -30,6 +32,12 @@ function detectContentType (ref, chunk) {
   return mime.contentType(mimeType)
 }
 
+async function resolveIpns (ref, ipfs) {
+  const [ root ] = PathUtils.splitPath(ref)
+  const immutableRoot = await ipfs.name.resolve(root, { recursive: true })
+  return ref.replace(`/ipns/${root}`, PathUtils.removeTrailingSlash(immutableRoot))
+}
```
[the remark this answers is not in the document]
Comment (lidel) [reply src/http/gateway/resources/gateway.js]: Good catch. I am pretty sure it is some old scar-tissue I used to get DNSLink work in Brave before your IPNS-related PRs landed. Removed it in  https://github.com/ipfs/js-ipfs/commit/7e3e7b262d1aea62dc6819931be1dc1c716487e0#diff-b8a2b51ac34c1e927c5129f0d0c019b2L35

File: src/http/gateway/resources/gateway.js [hunk 12]
```diff
@@ -45,29 +53,30 @@ class ResponseStream extends PassThrough {
...
-    const { ref } = request.pre.args
+    const ref = request.path
     const { ipfs } = request.server.app
 
+    // The resolver from ipfs-http-response supports only immutable /ipfs/ for now,
+    // so we convert /ipns/ to /ipfs/ before passing it to the resolver ¯\_(ツ)_/¯
+    // This could be removed if a solution proposed in
+    //  https://github.com/ipfs/js-ipfs-http-response/issues/22 lands upstream
+    const immutableRef = decodeURIComponent(ref.startsWith('/ipns/')
+      ? await resolveIpns(ref, ipfs)
```
[the remark this answers is not in the document]
Comment (lidel) [reply src/http/gateway/resources/gateway.js]: Fixed in 7e3e7b2

File: src/http/gateway/resources/gateway.js (hunk 8 above)
Comment (lidel) [reply src/http/gateway/resources/gateway.js]: done in f1da0014

File: src/http/gateway/resources/gateway.js (hunk 9 above)
Comment (lidel) [reply src/http/gateway/resources/gateway.js]: done in f1da0014

File: src/http/gateway/resources/gateway.js (hunk 10 above)
Comment (lidel) [reply src/http/gateway/resources/gateway.js]: done in f1da0014

File: src/http/gateway/resources/gateway.js [hunk 13]
```diff
@@ -45,29 +53,30 @@ class ResponseStream extends PassThrough {
...
-    if (!request.params.cid) {
-      throw Boom.badRequest('Path Resolve error: path must contain at least one component')
-    }
-
-    return { ref: `/ipfs/${request.params.cid}` }
-  },
 
   async handler (request, h) {
-    const { ref } = request.pre.args
+    const ref = request.path
```
[the remark this answers is not in the document]
Comment (lidel) [reply src/http/gateway/resources/gateway.js]: I looked at it and don't think we need to strip it, right?

[the remark this answers is not in the document]
Comment (hugomrdias) [reply src/http/gateway/resources/gateway.js]: nginx may add trailing slash thats my only issue

[the remark this answers is not in the document]
Comment (hugomrdias) [reply src/http/gateway/resources/gateway.js]: or even browsers

[the remark this answers is not in the document]
Comment (lidel) [reply src/http/gateway/resources/gateway.js]: I think that is already handled by code below that: 
- we add `/` for directories [here](https://github.com/ipfs/js-ipfs/blob/f1da001412087abe6e471485c39d361178bf4513/src/http/gateway/resources/gateway.js#L77-L80) 
- and remove it for files [here](https://github.com/ipfs/js-ipfs/blob/f1da001412087abe6e471485c39d361178bf4513/src/http/gateway/resources/gateway.js#L100-L103)

[the remark this answers is not in the document]
Comment (hugomrdias) [reply src/http/gateway/resources/gateway.js]: Cool ship it

File: src/http/api/routes/webui.js [hunk 14]
```diff
@@ -1,17 +1,26 @@
 'use strict'
 
+const Joi = require('@hapi/joi')
 const resources = require('../../gateway/resources')
 
 module.exports = [
   {
     method: '*',
-    path: '/ipfs/{cid*}',
+    path: '/ipfs/{path*}',
```
Comment (alanshaw) [inline src/http/api/routes/webui.js]: Does this need an `/ipns` handler also?

File: test/gateway/index.js [hunk 15]
```diff
@@ -84,6 +86,10 @@ describe('HTTP Gateway', function () {
       content('unsniffable-folder/hexagons-xml.svg'),
       content('unsniffable-folder/hexagons.svg')
     ])
+    // QmaRdtkDark8TgXPdDczwBneadyF44JvFGbrKLTkmTUhHk
+    await http.api._ipfs.add([content('utf8/cat-with-óąśśł-and-أعظم._.jpg')])
+    // Publish QmW2WQi7j6c7UgJTarActp7tDNikE4B2qXtFCfLPdsgaTQ to IPNS using self key
+    await http.api._ipfs.name.publish('QmW2WQi7j6c7UgJTarActp7tDNikE4B2qXtFCfLPdsgaTQ', { resolve: false })
```
Comment (alanshaw) [inline test/gateway/index.js]: Note for the future: unless we're trying to validate particular content creates a particular CID we should add the content and store the returned CID for use in future tests. Hard coding CIDs like this is a maintenance nightmare when we want to change defaults.

Comment (lidel) [reply test/gateway/index.js]: I am aware, but did not want to rewrite existing tests, so as a compromise I reused existing CID of `cat-folder/cat.jpg` (added few lines above)

File: src/http/api/routes/webui.js (hunk 14 above)
Comment (lidel) [reply src/http/api/routes/webui.js]: No, this is only for `/webui` on API port.
(`/ipns/` is only available on the Gateway port)

**Pull request merged**